tests/compile-and-dump/Singletons/T183.hs

```diff
@@ -0,0 +1,55 @@
+module T183 where
+
+import Data.Singletons.Prelude
+import Data.Singletons.TH
+
+$(singletons [d|
+  -----
+  -- Examples from #183
+  -----
+
+  f1 (x :: Maybe Bool) = (x :: Maybe Bool)
+  f2 (x :: Maybe a) = (x :: Maybe a)
+  f3 (Just a :: Maybe Bool) = "hi"
+
+  g x = case Just x of
+    (Just y :: Maybe Bool) -> (y :: Bool)
+
+  -----
+  -- Using explicit type signatures
+  -----
+
+  -- No explicit forall
+  foo1 :: Maybe a -> a
+  foo1 (Just x :: Maybe a) = (x :: a)
+
+  -- Explicit forall
+  foo2, foo3 :: forall a. Maybe a -> a
+  foo2 (Just x :: Maybe a) = (x :: a)
+  foo3 (Just x) = (x :: a)
+
+  -----
+  -- Multiple pattern signatures
+  -----
+
+  foo4 :: (a, b) -> (b, a)
+  foo4 = \(x :: a, y :: b) -> (y :: b, x :: a)
+
+  foo5, foo6 :: Maybe (Maybe a) -> Maybe (Maybe a)
+  foo5 (Just (Just (x :: a) :: Maybe a) :: Maybe (Maybe a))
+      = Just (Just (x :: a) :: Maybe a) :: Maybe (Maybe a)
+  foo6 (Just x :: Maybe (Maybe a))
+      = case x :: Maybe a of
+          (Just (y :: a) :: Maybe a) -> Just (Just (y :: a) :: Maybe a)
+
+  -----
+  -- Other pattern features
+  -----
+
+  foo7 :: a -> b -> a
+  foo7 (x :: a) (_ :: b) = (x :: a)
+
+  foo8 :: forall a. Maybe a -> Maybe a
+  foo8 x@(Just (_ :: a) :: Maybe a) = x
+  -- foo8 x@(Nothing :: Maybe a)       = x
```

Comment:
So it turns out that last line doesn't work at the moment, for a rather silly reason: it generates code like this:

```haskell
type family Let123X where
  Let123X = (NothingSym0 :: Maybe a)

type family Foo8 (x :: Maybe a) :: Maybe a where
  Foo8 (Nothing :: Maybe a) = Let123X

sFoo8 ::
  forall (t :: Maybe a).
  Sing t -> Sing (Foo8 t :: Maybe a)
sFoo8 SNothing
  = case SNothing of {
      (_ :: Sing (Nothing :: Maybe a))
        -> let
             sX :: Sing Let123X
             sX = SNothing :: Sing (NothingSym0 :: Maybe a)
           in sX }
```

Which doesn't typecheck:

```
../Bug.hs:14:3: error:
    • Could not deduce: a ~ a1
      from the context: t ~ 'Nothing
        bound by a pattern with constructor:
                   SNothing :: forall a. Sing 'Nothing,
                 in an equation for ‘sFoo8’
        at ../Bug.hs:(14,3)-(65,4)
      ‘a’ is a rigid type variable bound by
        the type signature for:
          sFoo8 :: forall a (t :: Maybe a). Sing t -> Sing (Apply Foo8Sym0 t)
        at ../Bug.hs:(14,3)-(65,4)
      ‘a1’ is a rigid type variable bound by
        the type signature for:
          sX :: forall a1. Sing Let6989586621679113062XSym0
        at ../Bug.hs:(14,3)-(65,4)
      Expected type: Sing Let6989586621679113062XSym0
        Actual type: Sing NothingSym0
    • In the expression: SNothing :: Sing (NothingSym0 :: Maybe a_aqR7)
      In an equation for ‘sX’:
          sX = SNothing :: Sing (NothingSym0 :: Maybe a_aqR7)
      In the expression:
        let
          sX :: Sing Let6989586621679113062XSym0
          sX = SNothing :: Sing (NothingSym0 :: Maybe a_aqR7)
        in sX
    • Relevant bindings include
        sX :: Sing Let6989586621679113062XSym0 (bound at ../Bug.hs:14:3)
        sFoo8 :: Sing t -> Sing (Apply Foo8Sym0 t)
          (bound at ../Bug.hs:14:3)
```

It turns out the culprit is the `sX :: Sing Let123X` type signature (commenting it out makes it typecheck). It looks like GHC can't figure out the `a` quantified by `sFoo8` should be the same `a1` that is an implicit argument to `Let123X`.

I would be that we could fix this by making that `a1` argument to `Let123X` visible and propagating that change through to its defunctionalization symbols, but I didn't have the energy to figure out how to do all of that in this patch. (It could probably be done separately, if need be.)

Reply:
Actually, it turns out that this is an entirely orthogonal issue, as you can trigger this problem without using pattern signatures. See https://github.com/goldfirere/singletons/issues/296.

Reply:
This is perhaps related to the problem I'm worried about above.